### ISSUE TITLE
RABBITMQ: FRAME_ERROR. [RE-OPEN]

### ISSUE THREAD

**Issue by adriancarayol** [pika/pika#882]
Hello everyone,

I have a problem with pika and rabbitmq, the user is connected to a channel, called "/profile/username"

When I send a 30MB file through a django form and ajax, I get the following error (I do not use channel or rabbitmq in the action of sending the form, (Django http request to a view)):

```
[2017/09/01 02:16:33] WebSocket CONNECT /profile/username/stream/ [127.0.0.1:45582]
WARNING:pika.adapters.base_connection:Socket closed when connection was open
WARNING:pika.connection:Disconnected from RabbitMQ at localhost:5672 (501): FRAME_ERROR - type 3, first 16 octets = <<42,171,223,145,112,216,92,19,214,170,218,214,186,
64,138,171>>: {invalid_frame_end_marker,39}
2017-09-01 02:16:39,792 - ERROR - server - Error trying to receive messages:
ERROR:daphne.server:Error trying to receive messages:
Exception in thread Thread-5:
Traceback (most recent call last):
File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
self.run()
File "/home/adrian/project/venv/lib/python3.5/site-packages/channels/management/commands/runserver.py", line 175, in run
worker.run()
File "/home/adrian/project/venv/lib/python3.5/site-packages/channels/worker.py", line 87, in run
channel, content = self.channel_layer.receive_many(channels, block=True)
File "/home/adrian/project/venv/lib/python3.5/site-packages/asgiref/base_layer.py", line 43, in receive_many
return self.receive(channels, block)
File "/home/adrian/project/venv/lib/python3.5/site-packages/asgi_rabbitmq/core.py", line 822, in receive
return future.result()
File "/usr/lib/python3.5/concurrent/futures/_base.py", line 405, in result
return self.__get_result()
File "/usr/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
raise self._exception
pika.exceptions.ConnectionClosed
Exception in thread Thread-3:
Traceback (most recent call last):
File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
self.run()
File "/home/adrian/project/venv/lib/python3.5/site-packages/channels/management/commands/runserver.py", line 175, in run
worker.run()
File "/home/adrian/project/venv/lib/python3.5/site-packages/channels/worker.py", line 87, in run
channel, content = self.channel_layer.receive_many(channels, block=True)
File "/home/adrian/project/venv/lib/python3.5/site-packages/asgiref/base_layer.py", line 43, in receive_many
return self.receive(channels, block)
File "/home/adrian/project/venv/lib/python3.5/site-packages/asgi_rabbitmq/core.py", line 822, in receive
return future.result()
File "/usr/lib/python3.5/concurrent/futures/_base.py", line 405, in result
return self.__get_result()
File "/usr/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
raise self._exception
pika.exceptions.ConnectionClosed

Exception in thread Thread-6:
Traceback (most recent call last):
File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
self.run()
File "/home/adrian/project/venv/lib/python3.5/site-packages/channels/management/commands/runserver.py", line 175, in run
worker.run()
File "/home/adrian/project/venv/lib/python3.5/site-packages/channels/worker.py", line 87, in run
channel, content = self.channel_layer.receive_many(channels, block=True)
File "/home/adrian/project/venv/lib/python3.5/site-packages/asgiref/base_layer.py", line 43, in receive_many
return self.receive(channels, block)
File "/home/adrian/project/venv/lib/python3.5/site-packages/asgi_rabbitmq/core.py", line 822, in receive
return future.result()
File "/usr/lib/python3.5/concurrent/futures/_base.py", line 405, in result
return self.__get_result()
File "/usr/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
raise self._exception
pika.exceptions.ConnectionClosed
INFO:neo4j.bolt:~~ [CLOSE]

Exception in thread Thread-4:
Traceback (most recent call last):
File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
self.run()
File "/home/adrian/project/venv/lib/python3.5/site-packages/channels/management/commands/runserver.py", line 175, in run
worker.run()
File "/home/adrian/project/venv/lib/python3.5/site-packages/channels/worker.py", line 87, in run
channel, content = self.channel_layer.receive_many(channels, block=True)
File "/home/adrian/project/venv/lib/python3.5/site-packages/asgiref/base_layer.py", line 43, in receive_many
return self.receive(channels, block)
File "/home/adrian/project/venv/lib/python3.5/site-packages/asgi_rabbitmq/core.py", line 822, in receive
return future.result()
File "/usr/lib/python3.5/concurrent/futures/_base.py", line 405, in result
return self.__get_result()
File "/usr/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
raise self._exception
pika.exceptions.ConnectionClosed
```

Software used:

- Pika 0.10.0
- Django 1.11.4
- Python3.5
- asgi_rabbitmq==0.5.3
- channels (last version)

I have created a project with the configuration of the project:

https://github.com/adriancarayol/issue-871

Steps:

install requeriments.txt
makemigrations + migrate
runserver
enter the URL: 127.0.0.1:8000/example
Upload a file larger than 30 MB
in the rabbitmq log file I also found this:

```
=ERROR REPORT==== 6-Sep-2017::00:34:55 ===
Error on AMQP connection <0.2183.0> (127.0.0.1:36466 -> 127.0.0.1:5672, vhost: '/', user: 'guest', state: running), channel 2:
operation none caused a connection exception frame_error: "type 3, first 16 octets = «132,2,158,13,120,86,89,177,63,47,150,109,77,252,\n                            102,160»: {invalid_frame_end_marker,83}"

=ERROR REPORT==== 6-Sep-2017::00:34:58 ===
closing AMQP connection <0.2183.0> (127.0.0.1:36466 -> 127.0.0.1:5672):
fatal_frame_error
```
Wireshark captures:
http://imgur.com/fxoKunol.png
http://imgur.com/dyjfivll.png
http://imgur.com/R3QYOS8l.png

Thanks.

**Comment by lukebakken** [comment]
@adriancarayol - the project you provided does not use the Pika module. I am unable to continue:

```
$ git remote -v
origin  https://github.com/adriancarayol/issue-871 (fetch)
origin  https://github.com/adriancarayol/issue-871 (push)

$ git grep -i pika
requeriments.txt:2:pika==0.11.0
```

**Comment by adriancarayol** [comment]
@lukebakken - I already use pika in the project, through "asgi_rabbitmq" package.

**Comment by lukebakken** [comment]
@adriancarayol I have attached a transcript of trying to reproduce the issue. When I browse to `localhost:8000/example`, I get the following exception:

```
2017-10-10 20:28:37,338 - INFO - server - Listening on endpoint tcp:port=8000:interface=127.0.0.1
Exception in thread Thread-3:
Traceback (most recent call last):
  File "/home/lbakken/.pyenv/versions/2.7.14/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/home/lbakken/.pyenv/versions/2.7.14/lib/python2.7/site-packages/channels/management/commands/runserver.py", line 176, in run
    worker.run()
  File "/home/lbakken/.pyenv/versions/2.7.14/lib/python2.7/site-packages/channels/worker.py", line 87, in run
    channel, content = self.channel_layer.receive_many(channels, block=True)
  File "/home/lbakken/.pyenv/versions/2.7.14/lib/python2.7/site-packages/asgiref/base_layer.py", line 43, in receive_many
    return self.receive(channels, block)
  File "/home/lbakken/.pyenv/versions/2.7.14/lib/python2.7/site-packages/asgi_rabbitmq/core.py", line 822, in receive
    return future.result()
  File "/home/lbakken/.pyenv/versions/2.7.14/lib/python2.7/site-packages/concurrent/futures/_base.py", line 429, in result
    return self.__get_result()
  File "/home/lbakken/.pyenv/versions/2.7.14/lib/python2.7/site-packages/concurrent/futures/_base.py", line 381, in __get_result
    raise exception_type, self._exception, self._traceback
ValueError: Must have completion callback with nowait=False
```
[transcript.txt](https://github.com/pika/pika/files/1373353/transcript.txt)

**Comment by lukebakken** [comment]
@adriancarayol if I use Pika `0.10.0` your example works. I suspect there is a change in `0.11.0` that `asgi_rabbitmq` is not handling. I'll see what I can do.

**Comment by adriancarayol** [comment]
@lukebakken my error is the same as this: http://john.eckersberg.com/debugging-rabbitmq-frame_too_large-error.html

**Comment by adriancarayol** [comment]
@lukebakken same at: @ #349

**Comment by lukebakken** [comment]
That _may_ be the same cause. Framing errors can have many different causes.

**Comment by proofit404** [comment]
Hi, guys.

I'm `asgi_rabbitmq` developer. I will try to upgrade my library to recent pika and check it against given project.

Hope I will have time to do it soon.

**Comment by kfrendrich** [comment]
Hi
I have the same problem with pika. I tried @adriancarayol 's example and got the some error.

```
2017-10-11 16:41:48,367 - ERROR - server - Error trying to receive messages: 
Exception in thread Thread-6:
Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/home/kfrendrich/codes/issue/lib/python3.5/site-packages/channels/management/commands/runserver.py", line 176, in run
    worker.run()
  File "/home/kfrendrich/codes/issue/lib/python3.5/site-packages/channels/worker.py", line 87, in run
    channel, content = self.channel_layer.receive_many(channels, block=True)
  File "/home/kfrendrich/codes/issue/lib/python3.5/site-packages/asgiref/base_layer.py", line 43, in receive_many
    return self.receive(channels, block)
  File "/home/kfrendrich/codes/issue/lib/python3.5/site-packages/asgi_rabbitmq/core.py", line 822, in receive
    return future.result()
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 405, in result
    return self.__get_result()
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
    raise self._exception
pika.exceptions.ConnectionClosed
Exception in thread Thread-5:
Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/home/kfrendrich/codes/issue/lib/python3.5/site-packages/channels/management/commands/runserver.py", line 176, in run
    worker.run()
  File "/home/kfrendrich/codes/issue/lib/python3.5/site-packages/channels/worker.py", line 87, in run
    channel, content = self.channel_layer.receive_many(channels, block=True)
  File "/home/kfrendrich/codes/issue/lib/python3.5/site-packages/asgiref/base_layer.py", line 43, in receive_many
    return self.receive(channels, block)
  File "/home/kfrendrich/codes/issue/lib/python3.5/site-packages/asgi_rabbitmq/core.py", line 822, in receive
    return future.result()
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 405, in result
    return self.__get_result()
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
    raise self._exception
pika.exceptions.ConnectionClosed


Exception in thread Thread-4:
Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/home/kfrendrich/codes/issue/lib/python3.5/site-packages/channels/management/commands/runserver.py", line 176, in run
    worker.run()
  File "/home/kfrendrich/codes/issue/lib/python3.5/site-packages/channels/worker.py", line 87, in run
    channel, content = self.channel_layer.receive_many(channels, block=True)
  File "/home/kfrendrich/codes/issue/lib/python3.5/site-packages/asgiref/base_layer.py", line 43, in receive_many
    return self.receive(channels, block)
  File "/home/kfrendrich/codes/issue/lib/python3.5/site-packages/asgi_rabbitmq/core.py", line 822, in receive
    return future.result()
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 405, in result
    return self.__get_result()
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
    raise self._exception
pika.exceptions.ConnectionClosed

Exception in thread Thread-3:
Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/home/kfrendrich/codes/issue/lib/python3.5/site-packages/channels/management/commands/runserver.py", line 176, in run
    worker.run()
  File "/home/kfrendrich/codes/issue/lib/python3.5/site-packages/channels/worker.py", line 87, in run
    channel, content = self.channel_layer.receive_many(channels, block=True)
  File "/home/kfrendrich/codes/issue/lib/python3.5/site-packages/asgiref/base_layer.py", line 43, in receive_many
    return self.receive(channels, block)
  File "/home/kfrendrich/codes/issue/lib/python3.5/site-packages/asgi_rabbitmq/core.py", line 822, in receive
    return future.result()
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 405, in result
    return self.__get_result()
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
    raise self._exception
pika.exceptions.ConnectionClosed

2017-10-11 16:41:53,368 - ERROR - server - Error trying to receive messages: 
2017-10-11 16:41:58,370 - ERROR - server - Error trying to receive messages: 
2017-10-11 16:42:03,372 - ERROR - server - Error trying to receive messages: 
2017-10-11 16:42:08,375 - ERROR - server - Error trying to receive messages: 
2017-10-11 16:42:13,376 - ERROR - server - Error trying to receive messages: 
2017-10-11 16:42:18,378 - ERROR - server - Error trying to receive messages: 
2017-10-11 16:42:23,380 - ERROR - server - Error trying to receive messages: 
2017-10-11 16:42:28,381 - ERROR - server - Error trying to receive messages: 
2017-10-11 16:42:33,384 - ERROR - server - Error trying to receive messages: 
2017-10-11 16:42:38,386 - ERROR - server - Error trying to receive messages: 
```

**Comment by lukebakken** [comment]
@kfrendrich can you provide code to reproduce this? Can you reprodue this using Pika version `0.11.0`?

**Comment by kfrendrich** [comment]
@lukebakken  thanks.
I used the same code that @adriancarayol posted, because it is hard to minify our application to demostrate it.

asgi-rabbitmq (0.5.3)
asgiref (1.1.2)
channels (1.1.8)
Django (1.11.6)
pika (0.10.0)

with pika 0.11.0 I experienced the same problem:
``` ValueError: Must have completion callback with nowait=False```

**Comment by lukebakken** [comment]
@kfrendrich - that is not the same problem. That `ValueError` is due to a change in the Pika API that `asgi_rabbitmq` must take into account.

@proofit404 I am going to open a PR to integrate Pika `0.11.0` into your project sometime today - have you also started work on that?

**Comment by adriancarayol** [comment]
@kfrendrich, @lukebakken, problem solved for me if use: https://github.com/adriancarayol/asgi_rabbitmq/tree/patch-1

**Comment by lukebakken** [comment]
@adriancarayol - when you say "problem solved" does that mean the "frame error" is solved as well?

I've made the same change and am running the test suite now. @proofit404 - PR on the way.

**Comment by adriancarayol** [comment]
@lukebakken yes, frame error solved, we are waiting for profit404 to do the pull request.

**Comment by kfrendrich** [comment]
Thank you guys. ;)
I am going to test it tomorrow morning.

**Comment by kfrendrich** [comment]
@lukebakken @adriancarayol  the upgrade solved the problems. Thank you.

**Comment by lukebakken** [comment]
Resolved by proofit404/asgi_rabbitmq#14
